Add blog post on shell operator (#3350)
tenzir/tenzir: web/blog/shelling-out-to-zeek-and-suricata/index.md

  element at index `1`, treats that as object and goes to the nested field
  `eve-log.filename`. Setting `/dev/stdout` as filename makes Suricata write to
  stdout.
- We must set `logging.outputs.0.console.enabled` to `no` because Suricata
  writes startup log messages to stdout. Since they are not valid JSON, we
  would otherwise create an invalid JSON output stream.

## User-defined Operators

Now that we have both Zeek and Suricata at our fingertips, how can we work with
their output more easily? This is where Tenzir comes into play—easy
[pipelines](/language/pipelines) for security teams to acquire,
[reshape](/user-guides/reshape-data), and route event data.

Here are two examples that count the number of unique source IP addresses per
destination IP address, on both Zeek and Suricata data:

```bash
# Zeek
zcat pcap.gz | zeekify | tenzir \
  'read zeek-json
   | where #schema == "zeek.conn"
   | summarize n=count_distinct(id.orig_h) by id.resp_h
   | sort n desc'
# Suricata
zcat pcap.gz | suricatify | tenzir \
  'read suricata
   | where #schema == "suricata.flow"
   | summarize n=count_distinct(src_ip) by dst_ip
   | sort n desc'
```

> It's a bit unwiedly to write such a command line that requires an external shell

Reviewer: ```suggestion
It's a bit unwieldy to write such a command line that requires an external shell
```

script to work. This is where [user-defined operators](/operators/user-defined)

> come into play. In combination with the [`shell`](/operators/sources/shell)
> operator in more detail, you can write a custom `zeek` and `suricata` operator
> and ditch the shell script:

Reviewer: ```suggestion
come into play. In combination with the [`shell`](/operators/sources/shell)
operator, you can write a custom `zeek` and `suricata` operator
and ditch the shell script:
```

```yaml title="tenzir.yaml"
tenzir:
  operators:
    zeek: >
     shell "zeek -r - LogAscii::output_to_stdout=T
            JSONStreaming::disable_default_logs=T
            JSONStreaming::enable_log_rotation=F
            json-streaming-logs"

>      | parse zeek-json

Reviewer: Opening this conversation here to remind us of the `parse -> read` change, which is scheduled to happen before this post is published.

Reviewer 2: Since that PR was just merged we need to replace parse with read in this blog post before merging.

    suricata: >
     shell "suricata -r /dev/stdin
            --set outputs.1.eve-log.filename=/dev/stdout
            --set logging.outputs.0.console.enabled=no"
     | parse suricata
```

The difference stands out when you look now at the pipeline definition:

```text title=Zeek
zeek
| where #schema == "zeek.conn"
| summarize n=count_distinct(id.orig_h) by id.resp_h
| sort n desc
```

```text bash title=Suricata
suricata
| where #schema == "suricata.flow"
| summarize n=count_distinct(src_ip) by dst_ip
| sort n desc
```

It's pretty convenient to drop packets into a Tenzir pipeline, process them with
our favorite tools, and then perform fast in-situ analytics on them. The nice
thing is that operators compose: a new operator automatically works with all
existing ones.

## How does it work?

First, let's take a look at the standard approach where one process pipes the
output into the next:

![Piping Zeek to Tenzir](zeek-to-tenzir-pipe.excalidraw.svg)

> When using the `shell` operator, the `tenzir` process spawns `zeek` or
> `suricata` as child process. The operator then copies the bytes from stdin of
> the `tenzir` process to the child's stdin using a Unix pipe (and in reverse for
> stdout):

Reviewer: ```suggestion
When using the `shell` operator, the `tenzir` process spawns `zeek` or
`suricata` as child process. The operator then forwards the bytes from stdin of
the `tenzir` process to the child's stdin, and uses the child's stdout as the input 
to the subsequent `parse` operator.
```

The changes to "copies" and "using a Unix pipe" (which I think is irrelevant) are just opinionated, but "and in reverse for stdout" is objectively incorrect.
